Sources/Private/CoreAnimation/Animations/ShapeAnimation.swift

```diff
@@ -120,4 +143,82 @@ extension Trim {
 
     return true
   }
+
+  /// Adjusts a `KeyframeGroup` associated with a `strokeStart` or `strokeEnd` to account
+  /// for trim offsets. The CoreAnimation rendering engine does not fully support trim offsets. If applying
+  /// the offsets results in a `KeyframeGroup` with all values in the range `[0, 1]` then the offsets
+  /// can be supported.
+  private func adjustKeyframesForTrimOffsets(
+    strokeKeyframes: ContiguousArray<Keyframe<Vector1D>>,
+    offsetKeyframes: ContiguousArray<Keyframe<Vector1D>>,
+    context _: LayerAnimationContext) throws -> ContiguousArray<Keyframe<Vector1D>> {
```

Comment:
Interesting, so this is basically combining the `trim.strokeStart` / `trim.strokeEnd` keyframes with the `offset` keyframes? Good to know. Could you add some discussion about this as a doc comment (discussion about why we have to combine these two values into a single set of keyframes would be super valuable!)

Reply:
I updated the comment but let me know if it still isn't clear. 

It should be possible to add full support for trim offsets. The main problem now is that an adjusted value `(strokeStart: 0.8, strokeEnd: 1.2)` cannot but supported because it crosses the "zero-point" of the animation. It should be possible to support this by appending the path to itself (creating two overlapping copies), and then adjusting that value again to be `(strokeStart: 0.4, strokeEnd: 0.6)` which is now valid and draws what was originally intended. 

This works in plain Core Animation code as implemented in [this Gist](https://gist.github.com/jtbergman/c138ddf3cf4082b261770b4d877c73b0), but when I tried to implement it [for Lottie](https://github.com/airbnb/lottie-ios/compare/master...jtbergman:jt-trim-offsets-full-support) it didn't work as expected. It seems the `CAShapeLayer` the stroke values are applied to is not the same `CAShapeLayer` the path is defined on? I'm also not sure the approach to duplicating the path works correctly in Lottie. Let me know if you have any ideas!

Reply:
I have an implementation of that path-duplicating approach here https://github.com/airbnb/lottie-ios/commit/416856c2f784025e85bf61d740eeb6815a174b44. This seems to sort-of-work, but isn't quite working with this `LottieFiles/loading_gradient_strokes` animation. I can't quite tell why, maybe `strokeStart` is growing to be larger than `strokeEnd` for a portion of the animation, causing the stroke to disappear entirely?

 | Default | Swapping `strokeStart` and `strokeEnd` keyframes (you can see the opposite portion of the animation works) |
| --- | ---- |
| ![2022-06-16 13 03 56](https://user-images.githubusercontent.com/1811727/174155330-0bebee33-c516-4b7d-af70-c09ff5b61ce9.gif) | ![2022-06-16 13 04 36](https://user-images.githubusercontent.com/1811727/174155362-7efd0697-28f4-4f17-b447-a344685270a0.gif) |

Reply:
It's clearly wrapping around the zero-point of the circle, though, which is progress! Would be amazing if we can figure out how to get this animation working.

Reply:
Nice work and thanks for doing this! 

> I can't quite tell why, maybe `strokeStart` is growing to be larger than `strokeEnd` for a portion of the animation, causing the stroke to disappear entirely?

I think this is the problem and it's caused by a bug that affects the original PR as well. In the original PR I just used the `inTangent` and `outTangent` of the most recent stroke or offset keyframe when creating the adjusted keyframes ([source](https://github.com/airbnb/lottie-ios/blob/1ce5ac6e9e032800cfb414ca7f687447698e745f/Sources/Private/CoreAnimation/Animations/ShapeAnimation.swift#L216-L219)). But the stroke and offset values can have independent `inTangent`/`outTangent` values which means the current approach to combining the keyframes is too simple, and I doubt there's an easy solution. As a next step, I'm going to try to manually interpolate the keyframes so that the `inTangent` and `outTangent` can be set to `nil`. My guess is this will be too expensive to use as-is, but it'll help verify the problem and approach.

Reply:
I implemented this and the performance seems to be ok! I need to touch up the code before pushing a new commit, but here's the result: 

| Main Thread | Core Animation |
| ----- | ------------- |
| <video src="https://user-images.githubusercontent.com/61445278/174523861-bed1810a-fced-4c9a-b47f-14850e093a82.mp4"> | <video src="https://user-images.githubusercontent.com/61445278/174523899-47aa4ac8-c703-4698-8a0f-d38e1fa27a50.mp4"> |

The animation for type-face G is also much better. 

| Master | Pull Request |
| ----- | ------------- |
| <video src="https://user-images.githubusercontent.com/61445278/174524198-7509969d-a1e3-4fff-9190-97b59f0bf358.mp4"> | <video src="https://user-images.githubusercontent.com/61445278/174523994-5e35da8f-aeb9-442e-b031-8d48672285ec.mp4"> |

Reply:
Amazing work!!!! Great stuff, thanks for the help